### PR TITLE
Add more tests for RGBASM code coverage

### LINKED ISSUE #1229: Add tests to improve code coverage
Follow-up to #996, lots of particular test gaps discovered by #1221.

Covered by #1257:

- [x] Test more invalid `STRFMT` format strings.
- [x] Test expanding macro args or interpolated symbols which contain `\n` `\r` `\t` characters.
- [x] Test illegal character escapes.
- [x] Test line continuations in string literals.
- [x] Try defining a macro inside a `REPT`, and invoke it inside+outside.
- [x] Test `BREAK` outside a `REPT`.
- [x] Test too-long, non-numeric, zero, and undefined `\<symbol>`.
- [x] Test invalid characters after line continuations.
- [x] Test `ELIF` after `ELSE` block.
- [x] Test invalid `OPT` values (also from the command line).
- [x] Test `PURGE` on an undefined symbol.
- [x] Test `EXPORT` on an anonymous label.
- [x] Test nested `REPT`/`FOR` blocks.
- [x] Test `BANK()` of a symbol in a fixed section.
- [x] Test defining data in RAM section.
- [x] Test impossible alignment in a fixed `SECTION UNION`.
- [x] Test impossible alignment in a fixed `SECTION FRAGMENT`.
- [x] Test `BANK()` for WRAM0 or HRAM symbols.
- [x] Test `SECTION` inside `UNION`.
- [x] Test `SECTION` inside `LOAD`.
- [x] Test `LOAD` inside `LOAD`.
- [x] Test `UNION` outside a section.
- [x] Test `UNION` inside `UNION`.
- [x] Test `NEXTU` and `ENDU` outside `UNION`.
- [x] Test unterminated `UNION`.
- [x] Test `ENDSECTION` outside a `SECTION`, and within a `UNION` or `LOAD`.

Covered by #1256:

- [x] `charmap_Delete` is an unused function.
- [x] Test `ATAN2`.
- [x] Test invalid fixed-point constant, e.g. `3.14q`.
- [x] Test integer constants that are too large.
- [x] Test no-digit `$` and `` ` `` constants.
- [x] Test an .asm file with CR-LF newlines.
- [x] Test macro args in string literals.
- [x] Test the `>=` operator.
- [x] Test `/` in a macro arg that isn't part of `/*`.
- [x] `macro_FreeArgs` is an unused function.
- [x] Test more than 4 `STRFMT` args so it reallocates.
- [x] Test more than 2 `PURGE` args so it reallocates.
- [x] Test `ASSERT FATAL`.
- [x] Test `FAIL`.
- [x] Test diff marks at the beginning of lines.
- [x] Test unary `+`.
- [x] Test `ldi` and `ldd`.
- [x] Test `ld [hl], [hl]`.
- [x] Test `ld` without `a`.
- [x] Test `rst` with an invalid address.
- [x] Test modulo by zero.
- [x] Test exponentiation by negative.
- [x] `sect_String` is an unused function.
- [x] `sym_FindUnscopedSymbol` is an unused function.
- [x] Test `@` outside a section.
- [x] Test `REDEF EQU` with a not-yet-defined name.
- [x] Test `REDEF EQU` on a built-in symbol.

### PR DESCRIPTION
**Pull request by Rangi42**
Fixes #1229

This doesn't provide 100% coverage for RGBASM, but a lot of what's left is just error checking for things we don't control (malloc failure, fopen failure, etc).

### CODE REVIEW

**Comment by ISSOtm** [comment]
The rationale for the old logic is that bits beyond the 16th don't exist, and thus are (mathematically) zeros—so only the lower 16 matter, hence the cap.

I do not like that the new logic silently discards constraints. I'd be okay with making non-satisfiable alignment constraints into errors, even for `align[17]` and beyond; the > 16 check was largely a path of least resistance.

**Comment by Rangi42** [comment]
What I want to avoid is a redundant error message, where first I get a warning my `ALIGN[20]` is capped to 16, and then the cap of 16 that the assembler chose itself causes another "unattainable" warning.

I considered printing one warning and capping to the maximum attainable value for a section (so ROM0 would still cap to 16, but ROMX would cap to 14), but that seemed too much like guessing what the user wants.

**Comment by ISSOtm** [comment]
I understand your sentiment, but I'm going to argue that it's a necessary evil.

First off, we have to cap the alignment size at some point, if only because of the data types used to represent it. And, given that, I think it makes more sense to cap at 16, considering that value's special significance.

Then, I agree two error messages is not great UX, but ultimately they're complementary: the latter's error message mentions the truncated alignment constraint, which would confuse anyone having specified too big an alignment constraint.

I think a better fix would be to run a check for "reachability", and suppressing the "over-aligning" warning; but I'm afraid that will regress in the presence of #524. Alternatively, rewording the "over-aligning" message to "addresses are 16-bit on the Game Boy, clamping alignment to 16" may better highlight its different nature.

**Comment by Rangi42** [comment]
> First off, we have to cap the alignment size at some point, if only because of the data types used to represent it. And, given that, I think it makes more sense to cap at 16, considering that value's special significance.

I don't think we do. When you write `SECTION "test", ROM0, ALIGN[20]`, the *parser* warns about the >16 alignment. So if the `ALIGN[20]` were ignored and treated as 0, it would be fine. Meanwhile, if you do `ALIGN[15]`, it will pass the parser and then `getSection` is free to complain about it being unattainable and, again, ignore it and treat it as 0 (which it already does).

Agreed that we can't let values >16 through, but setting them to 0 and erroring works as well as capping them at 16 for that.

**Comment by Rangi42** [comment]
(This ties into ax6's earlier request for out-of-range alignment values. `ALIGN 20` is an error that's ignored and doesn't affect the current alignment -- it's not capped as `ALIGN 16` --so IMO `SECTION ALIGN[20]` should be too.)

**Comment by ISSOtm** [comment]
> When you write `SECTION "test", ROM0, ALIGN[20]`, the _parser_ warns about the >16 alignment. So if the `ALIGN[20]` were ignored and treated as 0, it would be fine.

It's not, because that would mean that `HRAM, ALIGN[15]` is an error but `HRAM, ALIGN[16]` is not.

**Comment by ISSOtm** [comment]
> (This ties into ax6's earlier request for out-of-range alignment values. `ALIGN 20` is an error that's ignored and doesn't affect the current alignment -- it's not capped as `ALIGN 16` --so IMO `SECTION ALIGN[20]` should be too.)

If it's not capped, we have UB. The alignment is used as a LSH's RHS.

**Comment by Rangi42** [comment]
> It's not, because that would mean that `HRAM, ALIGN[15]` is an error but `HRAM, ALIGN[16]` is not.

No, they'd both be errors.
- `HRAM, ALIGN[17]` would make the parser complain "Alignment must be between 0 and 16".
- `HRAM, ALIGN[16]` and `HRAM, ALIGN[15]` are valid align specs, but then they'd make `getSection` complain "alignment cannot be attained".
- In both cases, an invalid alignment for that section type would cause a non-fatal error and would not affect the section's alignment.

> If it's not capped, we have UB. The alignment is used as a LSH's RHS.

True, although this was actually already being done -- in master there's a check for `if (alignOffset >= 1 << alignment) { error("Alignment offset (%" PRIu16 ") must be smaller than alignment size (%u)\n", alignOffset, 1U << alignment); alignOffset = 0; }` *before* the cap check for `if (alignment > 16) { error("Alignment must be between 0 and 16, not %u\n", alignment); alignment = 16; }`.
We'll just have to move the "is alignment > 16" check before both of those, and IMO make it ignore with `alignment = 0` instead of cap with `alignment = 16`.

File: src/asm/section.cpp [hunk 1]
```diff
@@ -303,6 +304,8 @@ static struct Section *createSection(char const *name, enum SectionType type,
 static struct Section *getSection(char const *name, enum SectionType type, uint32_t org,
 				  struct SectionSpec const *attrs, enum SectionModifier mod)
 {
+	assert(alignment < 16 && alignOffset < 1 << alignment); // Should be ensured by the caller
```
Comment (Rangi42) [inline src/asm/section.cpp:307]: `getSection` is only called by `sect_NewSection` and `sect_SetLoadSection`, and both of those are only called in parser.y, after parsing a `sectattrs`, which may have parsed an `align_spec`, which does its own bounds checking and error-messaging. (A couple other functions in section.cpp already have this assertion.)

File: src/asm/section.cpp [hunk 2]
```diff
@@ -307,6 +308,8 @@ static struct Section *getSection(char const *name, enum SectionType type, uint3
 	uint8_t alignment = attrs->alignment;
 	uint16_t alignOffset = attrs->alignOfs;
 
+	assert(alignment < 16 && alignOffset < 1 << alignment); // Should be ensured by the caller
```
Comment (Rangi42) [inline src/asm/section.cpp:311]: `getSection` is only called by `sect_NewSection` and `sect_SetLoadSection`, and both of those are only called in parser.y, after parsing a `sectattrs`, which may have parsed an `align_spec`, which does its own bounds checking and error-messaging. (A couple other functions in section.cpp already have this assertion.)

**Comment by Rangi42** [comment]
@ISSOtm I undid the change to section alignment warning behavior for the sake of getting all these tests merged. I do want to open a separate PR for it, because as you can see in test/asm/invalid-alignment.err, the current behavior gives redundant error messages.

**Review by ISSOtm — APPROVED** [review]
Surprising how much we took for granted. Thanks!

**Comment by ISSOtm** [comment]
> No, they'd both be errors.
> - `HRAM, ALIGN[17]` would make the parser complain "Alignment must be between 0 and 16".

I thought that was only a warning, not an error?

**Comment by Rangi42** [comment]
Thanks for reviewing all those!

> > * `HRAM, ALIGN[17]` would make the parser complain "Alignment must be between 0 and 16".
> 
> I thought that was only a warning, not an error?

Invalid alignment values have always been a non-fatal `error`. (It's what ax6 was requesting to change, making them an automatically-truncated warning like out-of-range `db`.)

**Pull request merged**